File: app/src/main/java/com/example/assignment/components/TitleApp.java
package com.example.assignment.components;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.res.TypedArray;
import android.util.AttributeSet;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.example.assignment.R;

import java.util.Objects;

public class TitleApp extends LinearLayout implements View.OnClickListener {
    static private final String TAG = TitleApp.class.getSimpleName();
    private TextView title, subTitle;
    private LinearLayout moreButton;
    private TitleButtonHandler titleButtonHandler;

    public TitleApp(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        LayoutInflater.from(context).inflate(R.layout.title_app_layout, this);
        TypedArray style = getContext().getTheme().obtainStyledAttributes(attrs, R.styleable.TitleApp, 0, 0);
        init(style);
    }

    public TitleApp(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        LayoutInflater.from(context).inflate(R.layout.title_app_layout, this);
        TypedArray style = getContext().getTheme().obtainStyledAttributes(attrs, R.styleable.TitleApp, 0, 0);
        init(style);
    }

    public TitleApp(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        LayoutInflater.from(context).inflate(R.layout.title_app_layout, this);
        TypedArray style = getContext().getTheme().obtainStyledAttributes(attrs, R.styleable.TitleApp, 0, 0);
        init(style);
    }

    public void init(TypedArray style) {
        title = findViewById(R.id.title);
        subTitle = findViewById(R.id.subTitle);
        moreButton = findViewById(R.id.moreButton);

        moreButton.setOnClickListener(TitleApp.this);

        //init style for component
        if (style == null) return;
        try {
            String textTitle = style.getString(R.styleable.TitleApp_app_title);
            String textSubTitle = style.getString(R.styleable.TitleApp_app_subTitle);
            boolean isMore = style.getBoolean(R.styleable.TitleApp_app_isMore, true);

            Log.d(TAG, "init: has handle: " + textSubTitle + " " + textTitle);

            //handle text title
            title.setText(textTitle != null ? textTitle : "");

            //handle text sub title
            if (textSubTitle != null) {
                subTitle.setText(textSubTitle);
            } else {
                subTitle.setVisibility(GONE);
            }

            //handle more button
            moreButton.setVisibility(isMore ? VISIBLE : GONE);

        } catch (Exception e) {
            Log.d(TAG, "init: " + e.getMessage());
        } finally {
            style.recycle();
        }
    }

    public void setTextTitle(@NonNull String text) {
        title.setText(text);
        requestLayout();
    }

    public void setTextSubTitle(@NonNull String text) {
        subTitle.setText(text);
        requestLayout();
    }

    public void setVisibleMoreButton(boolean visible) {
        moreButton.setVisibility(visible ? VISIBLE : GONE);
    }

    public void setTitleButtonHandler(TitleButtonHandler titleButtonHandler) {
        this.titleButtonHandler = titleButtonHandler;
        requestLayout();
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();
    }

    @Override
    public void onClick(View view) {
        if (view.getId() == R.id.moreButton) {
            if (titleButtonHandler != null) titleButtonHandler.onMoreButtonClick();
        }
    }

    public interface TitleButtonHandler {
        void onMoreButtonClick();
    }
}

File: app/src/main/java/com/example/assignment/components/TitleApp.kt
package com.example.assignment.components

import android.content.Context
import android.content.res.TypedArray
import android.util.AttributeSet
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.widget.LinearLayout
import android.widget.TextView
import com.example.assignment.R
import com.example.assignment.components.TitleApp

class TitleApp : LinearLayout, View.OnClickListener {
    private lateinit var title: TextView
    private var subTitle: TextView? = null
    private var moreButton: LinearLayout? = null
    private var titleButtonHandler: TitleButtonHandler? = null

    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs) {
        LayoutInflater.from(context).inflate(R.layout.title_app_layout, this)
        val style = getContext().theme.obtainStyledAttributes(attrs, R.styleable.TitleApp, 0, 0)
        init(style)
    }

    constructor(context: Context?, attrs: AttributeSet?, defStyleAttr: Int) : super(
        context,
        attrs,
        defStyleAttr
    ) {
        LayoutInflater.from(context).inflate(R.layout.title_app_layout, this)
        val style = getContext().theme.obtainStyledAttributes(attrs, R.styleable.TitleApp, 0, 0)
        init(style)
    }

    constructor(
        context: Context?,
        attrs: AttributeSet?,
        defStyleAttr: Int,
        defStyleRes: Int
    ) : super(context, attrs, defStyleAttr, defStyleRes) {
        LayoutInflater.from(context).inflate(R.layout.title_app_layout, this)
        val style = getContext().theme.obtainStyledAttributes(attrs, R.styleable.TitleApp, 0, 0)
        init(style)
    }

    fun init(style: TypedArray?) {
        title = findViewById(R.id.title)
        subTitle = findViewById(R.id.subTitle)
        moreButton = findViewById(R.id.moreButton)

        moreButton!!.setOnClickListener(this@TitleApp)

        //init style for component
        if (style == null) return
        try {
            val textTitle = style.getString(R.styleable.TitleApp_app_title)
            val textSubTitle = style.getString(R.styleable.TitleApp_app_subTitle)
            val isMore = style.getBoolean(R.styleable.TitleApp_app_isMore, true)
            Log.d(TAG, "init: has handle: $textSubTitle $textTitle")

            //handle text title
            title.text = textTitle ?: ""

            //handle text sub title
            if (textSubTitle != null) {
                subTitle!!.text = textSubTitle
            }
            subTitle!!.visibility = if (isMore) VISIBLE else GONE

            //handle more button
            moreButton!!.visibility = if (isMore) VISIBLE else GONE
        } catch (e: Exception) {
            Log.d(TAG, "init: " + e.message)
        } finally {
            style.recycle()
        }
    }

    fun setTextTitle(text: String) {
        title.text = text
        requestLayout()
    }

    fun setTextSubTitle(text: String) {
        subTitle!!.text = text
        requestLayout()
    }

    fun setVisibleMoreButton(visible: Boolean) {
        moreButton!!.visibility = if (visible) VISIBLE else GONE
    }

    fun setTitleButtonHandler(titleButtonHandler: TitleButtonHandler) {
        this.titleButtonHandler = titleButtonHandler
        requestLayout()
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
    }

    override fun onClick(view: View) {
        if (view.id == R.id.moreButton && titleButtonHandler != null) {
            titleButtonHandler!!.onMoreButtonClick()
        }
    }

    interface TitleButtonHandler {
        fun onMoreButtonClick()
    }

    companion object {
        private val TAG = TitleApp::class.java.simpleName
    }
}